pep-0497-externals.txt

```diff
@@ -0,0 +1,359 @@
+many minutes in to a large compile job.
+
+A key aspect of the usability of this tool is a mechanism to provide more useful
+information to the user about how to obtain those missing resources.
+The proposal is to provide a layer inside of pip which is capable of providing
+information about appropriate package names. It is surprising that vendors
+do not already provide a good tool for this, given that finding appropriate
+header and library files is a common experience in the general linux ecosystem.
+However, it is also a critical aspect of the user experience, and something
+which package maintainers will need to be able to specify precisely. The
+proposed approach would be for pip to supply the ability to install an
+inference layer, usable by users and vendors alike. Usage would be along
+the lines of
+
+    pip install vendor_magic darwin
+
+This would transform the user messages relating to missing dependencies from
+
+    Could not find libatlas.so
+
+to
+
+    Could not find libatlas.so. Run 'port install scipy +atlas'
+
+
+Discussion / FAQ / Issues
+=========================
+
+This section will be updated to capture incoming feedback and represents a
+snapshot of the current concerns.
+
+
+Specification of files (headers, libs, bins) vs packages
+--------------------------------------------------------
+
+There was a discussion about whether specific dependencies (named headers,
+libraries, binaries etc) should be included, or whether general package names
+(such as may be expected by yum, apt, conda) were appropriate. Due to the
+number of package management tools, plus the changing nature of those, it
+was decided that concentrating on the specific files was more useful than
+using package names. It also provides looser coupling between the specification
+of externals and the package management tools.
+
+Location in setup.py rather than requirements.txt
+-------------------------------------------------
+
+Requirements.txt is scoped to the use for local deployments to specify
+requirements precisely for the relevant local environment. Setup.py is the
+appropriate place to specify the information for the package in general.
+
+In future, there may be some scope for including similar directives within
+requirements.txt, but that is out of scope for this PEP.
+
+Prior Approaches
+----------------
+
+Some capacity to support description of external dependencies (among other
+things) was described in PEPs 314 and 345 (Metadata 1.1 and 1.2). These
+describe directives whichs can be placed in the PKG-INFO file. An implementation
+responding to these directives is pending.
+
+This PEP places the directives in setup.py. This approach is closer to the
+typical approach of package maintainers to expressing requirements and places
+the external dependency specification nearer to the Python dependency
+specification. This PEP would supercede PEPs 314 and 345 with respect to
+external dependencies.
```

Comment:
This answer isn't quite right. The problem with PKG-INFO (and metadata 1.x in general) is that it doesn't clearly distinguish "build time metadata passed from developers to their chosen build tool" from "install time metadata passed from a developer's chosen build tool to an end user's chosen install tool".

There's actually an open question for this PEP as to whether the design should be "developer specifies expected external dependencies in the input metadata, build tool may complain if the actual external dependencies go beyond the expected dependencies", or "build tool automatically adds external dependencies to the generated metadata".

I think the PEP gets it right (where external dependencies are only added deliberately, and we make the input metadata richer so we can compare intent and actual dependencies), but the rationale in this answer could be improved by considering the trade-off more explicitly.

Reply:
In conda-build, we have a detection step that examines the linkage of any binaries.  For dependencies that get detected there, we match them up with package names in conda where those libraries can be obtained.  If those packages are not part of our "run" dependencies, we warn, but conda-build can also be configured to error out.  This has dramatically improved the quality of our packages.

It's a lot harder to think about how to do this in the general case.  I suppose as long as you only compare your manually specified dependency against detected dependencies, that's pretty viable.  Comparing against arbitrary external package managers to ensure that some external package dependency is listed would be a whole lot harder, though.